### PR TITLE
[Feature] Support iceberg schema evolution in Parquet format

### PR DESCRIPTION
**Pull request by Smith-Cruise**
Signed-off-by: Smith Cruise <chendingchao1@126.com>

## What type of PR is this：
- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Support read parquet after iceberg schema evolution.
We use iceberg global unique field id to access column instead of column name.
## Which issues of this PR fixes ：

Fixes #

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [ ] I have checked the version labels which the pr will be auto backported to target branch
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3
  - [ ] 2.2

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 0 / 19 (00.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/IcebergTable.java|0|1|00.00%|[403]|
|:large_blue_circle:|com/starrocks/connector/iceberg/IcebergUtil.java|0|18|00.00%|[157, 158, 159, 160, 161, 162, 163, 164, 168, 169, 170, 171, 172, 173, 174, 175, 176, 178]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 0 / 19 (00.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/IcebergTable.java|0|1|00.00%|[403]|
|:large_blue_circle:|com/starrocks/connector/iceberg/IcebergUtil.java|0|18|00.00%|[157, 158, 159, 160, 161, 162, 163, 164, 168, 169, 170, 171, 172, 173, 174, 175, 176, 178]|

File: gensrc/thrift/Descriptors.thrift [hunk 1]
```diff
@@ -344,12 +344,33 @@ struct TFileTable {
     2: optional list<TColumn> columns
 }
 
+struct TIcebergSchema {
+    1: optional list<TIcebergSchemaField> fields
+}
+
+struct TIcebergSchemaField {
+    // Refer to field id in iceberg schema
+    1: optional i32 field_id
```
Comment (DorianZheng) [inline gensrc/thrift/Descriptors.thrift:353]: Seems it's better to add this in `TColumn`. So we don't have to define it twice.

Comment (Smith-Cruise) [reply gensrc/thrift/Descriptors.thrift:353]: `TColumn` is not only used for IcebergTable, I think it's better to put it in IcebergTable.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 0 / 19 (00.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/IcebergTable.java|0|1|00.00%|[403]|
|:large_blue_circle:|com/starrocks/connector/iceberg/IcebergUtil.java|0|18|00.00%|[157, 158, 159, 160, 161, 162, 163, 164, 168, 169, 170, 171, 172, 173, 174, 175, 176, 178]|

File: be/src/formats/parquet/file_reader.cpp [hunk 2]
```diff
@@ -42,18 +41,47 @@ Status FileReader::init(HdfsScannerContext* ctx) {
     _scanner_ctx = ctx;
     RETURN_IF_ERROR(_parse_footer());
 
+    // set existed SlotDescriptor in this parquet file
     std::unordered_set<std::string> names;
-    _file_metadata->schema().get_field_names(&names);
+    RETURN_IF_ERROR(_set_existed_column_names(&names));
```
Comment (dirtysalt) [inline be/src/formats/parquet/file_reader.cpp:46]: set or get?

File: be/src/formats/parquet/file_reader.cpp [hunk 3]
```diff
@@ -189,9 +217,40 @@ StatusOr<bool> FileReader::_filter_group(const tparquet::RowGroup& row_group) {
...
+                                              _file_metadata->schema().get_field_pos_by_field_id(it->second));
+        }
+    }
+
     for (size_t i = 0; i < slots.size(); i++) {
         const SlotDescriptor* slot = slots[i];
-        const auto* column_meta = _get_column_meta(row_group, slot->col_name(), _scanner_ctx->case_sensitive);
+        const tparquet::ColumnMetaData* column_meta = nullptr;
+        if (_scanner_ctx->iceberg_schema == nullptr) {
+            column_meta =
```
Comment (dirtysalt) [inline be/src/formats/parquet/file_reader.cpp:249]: there are a lot of if/else here? can we extract interface and gives them two implementations?

Comment (Smith-Cruise) [reply be/src/formats/parquet/file_reader.cpp:228]: improved

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 0 / 19 (00.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/IcebergTable.java|0|1|00.00%|[403]|
|:large_blue_circle:|com/starrocks/connector/iceberg/IcebergUtil.java|0|18|00.00%|[157, 158, 159, 160, 161, 162, 163, 164, 168, 169, 170, 171, 172, 173, 174, 175, 176, 178]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 0 / 19 (00.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/IcebergTable.java|0|1|00.00%|[403]|
|:large_blue_circle:|com/starrocks/connector/iceberg/IcebergUtil.java|0|18|00.00%|[157, 158, 159, 160, 161, 162, 163, 164, 168, 169, 170, 171, 172, 173, 174, 175, 176, 178]|

**Review by dirtysalt — DISMISSED** [review]

File: be/src/formats/parquet/meta_helper.cpp [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+                                                        materialized_column.col_type, materialized_column.slot_id);
+        read_cols.emplace_back(column);
+    }
+    _is_only_partition_scan = read_cols.empty();
+}
+
+void IcebergMetaHelper::set_existed_column_names(std::unordered_set<std::string>* names) const {
+    names->clear();
+    // build column name set from iceberg schema
+    auto fields = _t_iceberg_schema->fields;
```
Comment (stephen-shelby) [inline be/src/formats/parquet/meta_helper.cpp:68]: auto -> auto& ?

File: fe/fe-core/src/main/java/com/starrocks/connector/iceberg/IcebergUtil.java [hunk 5]
```diff
@@ -149,6 +153,31 @@ public static RemoteFileInputFormat getHdfsFileFormat(FileFormat format) {
         }
     }
 
+    public static TIcebergSchema getTIcebergSchema(Schema schema) {
+        Types.StructType rootType = schema.asStruct();
+        TIcebergSchema tIcebergSchema = new TIcebergSchema();
+        List<TIcebergSchemaField> fields = new LinkedList<>();
```
Comment (stephen-shelby) [inline fe/fe-core/src/main/java/com/starrocks/connector/iceberg/IcebergUtil.java:159]: nit: ArrayList

File: be/src/formats/parquet/meta_helper.cpp (hunk 4 above)
Comment (Smith-Cruise) [reply be/src/formats/parquet/meta_helper.cpp]: fixed

File: fe/fe-core/src/main/java/com/starrocks/connector/iceberg/IcebergUtil.java (hunk 5 above)
Comment (Smith-Cruise) [reply fe/fe-core/src/main/java/com/starrocks/connector/iceberg/IcebergUtil.java]: fixed

File: be/src/formats/parquet/schema.h [hunk 6]
```diff
@@ -63,6 +63,9 @@ struct ParquetField {
     // used to get ColumnChunk in parquet file's metadata
     int physical_column_index;
 
+    // Unique field id for schema evolution
```
Comment (stephen-shelby) [inline be/src/formats/parquet/schema.h:66]: field id in parquet schema? not only for schema evolution

File: be/src/formats/parquet/meta_helper.h [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+
+class IcebergMetaHelper : public MetaHelper {
+public:
+    IcebergMetaHelper(std::shared_ptr<FileMetaData> file_metadata, bool case_sensitive,
+                      const TIcebergSchema* t_iceberg_schema) {
+        _file_metadata = file_metadata;
+        _case_sensitive = case_sensitive;
+        _t_iceberg_schema = t_iceberg_schema;
+        DCHECK(_t_iceberg_schema != nullptr);
+    }
```
Comment (zombee0) [inline be/src/formats/parquet/meta_helper.h:116]: what about we update be first and then fe? we'd better give some prompt information here.

File: be/src/formats/parquet/meta_helper.h [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+public:
+    virtual ~MetaHelper() = default;
+    virtual void set_existed_column_names(std::unordered_set<std::string>* names) const = 0;
+
+    virtual void build_column_name_2_pos_in_meta(std::unordered_map<std::string, size_t>& column_name_2_pos_in_meta,
+                                                 const tparquet::RowGroup& row_group,
+                                                 const std::vector<SlotDescriptor*>& slots) const = 0;
+
+    virtual const void prepare_read_columns(const std::vector<HdfsScannerContext::ColumnInfo>& materialized_columns,
+                                            std::vector<GroupReaderParam::Column>& read_cols,
```
Comment (zombee0) [inline be/src/formats/parquet/meta_helper.h:39]: why did you use const void, could you confirm it?

File: be/src/formats/parquet/meta_helper.h (hunk 7 above)
Comment (zombee0) [reply be/src/formats/parquet/meta_helper.h:116]: if it go non-iceberg-schema branch, schema evolution not work, where can we give some message?

Comment (Smith-Cruise) [reply be/src/formats/parquet/meta_helper.h:116]: Just like before, query will return empty result.

File: be/src/formats/parquet/meta_helper.h (hunk 8 above)
Comment (Smith-Cruise) [reply be/src/formats/parquet/meta_helper.h:39]: I will change it.

File: be/src/formats/parquet/meta_helper.h [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+        auto it = column_name_2_pos_in_meta.find(col_name);
+        if (it == column_name_2_pos_in_meta.end()) {
+            return nullptr;
+        }
+
+        if (it->second >= row_group.columns.size()) {
+            DCHECK(false) << "Impossible here";
+            return nullptr;
+        }
+
```
[the remark this answers is not in the document]
Comment (Smith-Cruise) [reply be/src/formats/parquet/meta_helper.h:54]: But we need to return nullptr🤣, avoid array out of the index.

File: be/src/formats/parquet/meta_helper.h (hunk 7 above)
Comment (zombee0) [reply be/src/formats/parquet/meta_helper.h:116]: ok

File: be/src/formats/parquet/meta_helper.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (zombee0) [reply be/src/formats/parquet/meta_helper.h:54]: ok, it's not really impossible, haha

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 0 / 19 (00.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/IcebergTable.java|0|1|00.00%|[160]|
|:large_blue_circle:|com/starrocks/connector/iceberg/IcebergApiConverter.java|0|18|00.00%|[96, 97, 98, 99, 100, 101, 102, 103, 107, 108, 109, 110, 111, 112, 113, 114, 115, 117]|

File: be/src/formats/parquet/column_reader.cpp [hunk 10]
```diff
@@ -538,6 +527,95 @@ Status ColumnReader::do_create(const ColumnReaderOptions& opts, const ParquetFie
...
+        }
+
+        std::vector<std::unique_ptr<ColumnReader>> children_readers;
+        for (size_t i = 0; i < col_type.children.size(); i++) {
+            const auto& format_subfield_name = opts.case_sensitive
+                                                       ? col_type.field_names[i]
+                                                       : boost::algorithm::to_lower_copy(col_type.field_names[i]);
+
+            auto iceberg_it = subfield_name_2_field_id.find(format_subfield_name);
+            if (iceberg_it == subfield_name_2_field_id.end()) {
```
[the remark this answers is not in the document]
Comment (Smith-Cruise) [reply be/src/formats/parquet/column_reader.cpp]: Yes, that should not happend, this code is defensive.

File: be/src/formats/parquet/column_reader.cpp [hunk 11]
```diff
@@ -538,6 +527,95 @@ Status ColumnReader::do_create(const ColumnReaderOptions& opts, const ParquetFie
...
+        if (!col_type.children[1].is_unknown_type()) {
+            RETURN_IF_ERROR(ColumnReader::create(opts, &(field->children[0].children[1]), col_type.children[1],
+                                                 value_iceberg_schema, &value_reader));
+        }
+
+        std::unique_ptr<MapColumnReader> reader(new MapColumnReader(opts));
+        RETURN_IF_ERROR(reader->init(field, std::move(key_reader), std::move(value_reader)));
+        *output = std::move(reader);
+    } else if (field->type.type == LogicalType::TYPE_STRUCT) {
+        std::unordered_map<std::string, const TIcebergSchemaField*> subfield_name_2_field_id{};
```
[the remark this answers is not in the document]
Comment (Smith-Cruise) [reply be/src/formats/parquet/column_reader.cpp]: fixed

**Review by zombee0 — DISMISSED** [review]

**Review by Youngwb — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Review by zombee0 — APPROVED** [review]

**Review by Youngwb — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by Smith-Cruise** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 0 / 19 (00.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/IcebergTable.java|0|1|00.00%|[160]|
|:large_blue_circle:|com/starrocks/connector/iceberg/IcebergApiConverter.java|0|18|00.00%|[96, 97, 98, 99, 100, 101, 102, 103, 107, 108, 109, 110, 111, 112, 113, 114, 115, 117]|

**Comment by Smith-Cruise** [comment]
run starrocks_admit_test

**Comment by Smith-Cruise** [comment]
run starrocks_admit_test

**Comment by Smith-Cruise** [comment]
run starrocks_tscancode

**Review by dirtysalt — APPROVED** [review]

**Comment by Smith-Cruise** [comment]
@Mergify backport branch-3.0

**Pull request merged**